magento/magento2
# AC-2466 - Added ability to retrieve config paths instead of XML paths for use when re-encrypting data

### Description (*)

This PR aims to resolve issue #35061. The issue describes how data is not re-encrypted when a new encryption key is being generated.

I found that the `Change` controller retrieves all config fields from the XML structure based on the `backend_model` being set to `\Magento\Config\Model\Config\Backend\Encrypted`. While this works, the paths which are returned are the full XML paths. There is however an option to define a `<config_path>` node, which will be used in the `core_config_data` table to store the configuration value.

Since the Controller checks directly against this table, a lot of paths seem to be "incorrect" and are not retrieved as values to re-encrypt.

In this PR I added a parameter (for backward compatibility) to retrieve the config path, instead of the full XML path.

### Related Pull Requests

N/A

### Fixed Issues (if relevant)

1. Fixes magento/magento2#35061

### Manual testing scenarios (*)

1. Store an encrypted value where the XML contains a config path (eg. Stores > Configuration > Sales > Payment Methods > PayPal Express Checkout > API Password
2. Check the value of the encrypted configuration in the database (the first digit indicates the encryption key version). _Additionally you can use a tool like Magerun to verify the decrypted value (`config:store:get --decrypt paypal/wpp/api_password`)_
4. Navigate to the encryption configuration page ( System > Manage Encryption Key)
5. Generate a new encryption key (Auto-generate a Key > Yes > Click: Change Encryption Key)
6. After the encryption key has been generated, check if the encryption has been updated in the database. _Again, you can use a tool like Magerun to verify the decrypted value (`config:store:get --decrypt paypal/wpp/api_password`)_

### Questions or comments

1. If it's safe to do so, I would suggest always returning the config path if it's available, meaning the changes in: `\Magento\Config\Model\Config\Structure::_getGroupFieldPathsByAttribute` could be simplified to a simple: `$field['config_path'] ?? $parentPath . '/' . $field['id']`.
2. I've added strict types, but I'm not sure if this is already the standard for core Magento. if not this can be removed.

### Contribution checklist (*)
